### PR TITLE
Align KTO with DPO: Align tokenization

### PR DESCRIPTION
**Pull request by albertvillanova**
Align KTO with DPO: Align tokenization.

This PR refactors the tokenization logic in `KTOTrainer` to improve clarity, modularity, and support for conversational data. The main changes include removing the old `_tokenize` function, introducing a new instance method `_tokenize` that handles both conversational and non-conversational inputs, and updating dataset preparation to use this new logic. Additionally, the pull request simplifies and clarifies the token processing pipeline.

Part of:
- #4786

### Changes

**Tokenization Refactor and Conversational Data Support:**

* Removed the original `_tokenize` function, which was previously responsible for tokenizing batches, and replaced it with a new instance method `_tokenize` that can handle both plain text and conversational (chat) inputs, improving modularity and clarity. 
* Updated dataset preparation in `_prepare_dataset` to use the new `_tokenize` method and added support for conversational data via the `is_conversational` utility, ensuring correct tokenization for both chat and non-chat datasets. 

**Code Cleanup and Simplification:**

* Removed unused imports such as `numpy` and the type import for `PreTrainedTokenizer`, streamlining dependencies.
* Simplified the `_process_tokens` function by removing redundant checks and batch construction, focusing it on label processing.

---

> [!NOTE]
> **Medium Risk**
> Changes how training/eval datasets are tokenized (including chat-style inputs), which can subtly affect prompt/answer boundaries and thus loss computation; tests were updated to cover the new behavior but tokenizer edge cases may still surface.
> 
> **Overview**
> Refactors `KTOTrainer` dataset preprocessing to use a new instance `_tokenize` method and a per-example `tokenize_fn`, adding explicit handling for conversational inputs via `is_conversational`/`apply_chat_template` and emitting a warning when prompt vs prompt+completion tokenization mismatches.
> 
> Removes the old batch `_tokenize` implementation (and its `numpy` dependency) and simplifies `_process_tokens` to operate purely on pre-tokenized fields while preserving the `label`.
> 
> Updates `test_tokenize_and_process_tokens` to validate tokenization through `trainer.train_dataset` (rather than calling `_tokenize` directly), adds KL corruption checks against the prepared dataset, and adds a synthetic `Dataset.from_dict` case to exercise `_process_tokens` with `num_proc`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8780ca17ae373478ca3368fb05da076b6175f656. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/kto/kto_trainer.py [hunk 1]
```diff
@@ -642,14 +592,39 @@ def _prepare_dataset(
...
-                desc=f"Tokenizing {dataset_name} dataset",
-            )
+            def tokenize_fn(example, processing_class):
+                if is_conversational(example):
+                    prompt_ids = self._tokenize(processing_class, example["prompt"], add_generation_prompt=True)[
+                        "input_ids"
+                    ]
+                    prompt_completion_ids = self._tokenize(
+                        processing_class, example["prompt"] + example["completion"]
+                    )["input_ids"]
```
[the remark this answers is not in the document]
Comment (albertvillanova) [reply trl/experimental/kto/kto_trainer.py:604]: Yes, this is intentional. I am addressing this in my follow-up PR: Remove maybe_apply_chat_template

File: tests/experimental/test_kto_trainer.py [hunk 2]
```diff
@@ -15,11 +15,11 @@
 import multiprocess
 import pytest
 import torch
-from datasets import load_dataset
+from datasets import Dataset, load_dataset
 from transformers import AutoModelForCausalLM, AutoTokenizer
 
 from trl.experimental.kto import KTOConfig, KTOTrainer
-from trl.experimental.kto.kto_trainer import _get_kl_dataset, _process_tokens, _tokenize
```
Comment (qgallouedec) [inline tests/experimental/test_kto_trainer.py:22]: 👍

**Review by qgallouedec — APPROVED** [review]
lgtm!
